src/14_legendre.jl

```diff
@@ -0,0 +1,362 @@
+
+    ABC_recurrence(n, x,
+        ( ((2m-one(x))/m, zero(x), (m-one(x))/m) for m=2:n),            # (A_m, B_m, C_m)
+        Q0, Q1)
+end
+function legendreq(n::Integer, m::Integer, x)
+    (n <  0) && throw(DomainError(n, "must be non-negative"))
+    (m <  0) && throw(DomainError(m, "must be non-negative"))
+    (m == 0) && return legendreq(n, x)
+
+    # x check after: Q^0_n is also implemented for x = +-1
+    (1 <= abs(x)) && throw(DomainError(x, "must be in the range (-1,1)"))
+
+    # step 1: compute Q_n^0(x)
+    q_n_0 = legendreq(n, x)
+
+    # step 2: compute Q_n^1(x)
+    Q01 = -(1-x^2)^(-0.5)                                                   # = Q_0^1(x)
+    Q11 = -sqrt(1-x^2) * (0.5*log((1+x)/(1-x)) + x/(1-x^2))                 # = Q_1^1(x)
+    if n == 0
+        q = Q01
+    elseif n == 1
+        q = Q11
+    else
+        q, q_prev = Q11, Q01
+        for k = 2:n
+            q_prev, q_prev_prev = q, q_prev
+            q = (2k-1) / (k-1) * x * q_prev - k / (k-1) * q_prev_prev       # = Q_k^1(x)
+        end                                                                 # on output: q = Q_n^1(x)
+    end
+    q_n_1 = q
+
+    # step 3: compute Q_n^m(x)
+    q_prev  = q_n_0
+    q       = q_n_1
+    for k = 2:m
+        q_prev, q_prev_prev = q, q_prev
+        q = -2*(k-1)*x*q_prev / sqrt(1-x^2) - (n+k-1)*(n-k+2)*q_prev_prev   # = Q_n^k(x)
+    end                                                                     # on output: q = Q_n^m(x)
+
+    q
+end
+
+@doc raw"""
+    ABC_recurrence(n, x, ABC, p0, p1)
+
+Evaluate polynomials by the recurrence
+```math
+p_n(x)
+=
+\begin{cases}
+    p_0(x)
+    & \text{for} \quad n = 0
+    \\
+    p_1(x)
+    & \text{for} \quad n = 1
+    \\
+    (A_n x + B_n) p_{n-1}(x) - C_n p_{n-2}(x)
+    & \text{for} \quad n \geq 2
+    \,.
+\end{cases}
+```
+
+This will be used to compute Legendre ``P_n``, Laguerre ``L_n``, Hermite ``H_n``,
+and Chebyshev polynomials of 1st kind ``T_n`` and 2nd kind ``U_n``.
+
+# Arguments
+- `n::Integer`:                     Polynomial order
+- `x`:                              evaluation point
+- `ABC::Base.Generator`:            recurrence coefficients ``(A_m,B_m,C_m)`` for ``m=2,\dots,n``
+- `p0`:                             function of degree 0, i.e. ``p_0``, evaluated at ``x``
+- `p1`:                             function of degree 1, i.e. ``p_1``, evaluated at ``x``
+
+#Implementation
+The coefficients ``A_m, B_m, C_m`` are supplied via a generator which yields them as a tuple ``(A_m,B_m,C_m)``.
+
+Type stability:
+- The generator `ABC` yields tuples `(Am,Bm,Cm)`. The tuple's elements are of the same type as `x`.
+- `p0,p1` are functions which have the same return type as the type of `x`.
+- Thus, all variables: `p_prev_prev,p_prev,p,Am,Bm,Cm` are always of type `x`.
+
+# Reference
+> JIN, J. M.;
+> JJIE, Zhang Shan.
+> "Computation of special functions".
+> Wiley, 1996.
+> (p. 23)
+"""
+function ABC_recurrence(n::Integer, x::T,
```

Comment:
I appreciate the desire to abstract out the 3-term recurrence, but I think it's going to be hard to make this performant and generic without using a macro to do more explicit code generation.    If you're not up for macros, it's really not that much code to just implement the recurrence separately in each routine.

Reply:
I removed the recurrence function (see commit [780b6df](https://github.com/JuliaMath/SpecialFunctions.jl/pull/175/commits/780b6df504afa8169b6529f46b78dcd1b9350ef6)) and hardcoded each function's recurrence on its own. This is more flexible anyways and alllows for more efficient implementations (see comment by @MikaelSlevinsky above).